storage/txcache/evictionScoreComputer.go

```diff
@@ -0,0 +1,145 @@
+package txcache
+
+import "math"
+
+type evictionScoreComputer struct {
+	senders          []*txListForSender
+	scores           []float64
+	scoresAsPercents []int64
+
+	maxGas         int64
+	minGas         int64
+	maxSize        int64
+	minSize        int64
+	maxOrderNumber int64
+	minOrderNumber int64
+	maxTxCount     int64
+	minTxCount     int64
+
+	gasRange         int64
+	sizeRange        int64
+	orderNumberRange int64
+	txCountRange     int64
+}
+
+func newEvictionScoreComputer(senders []*txListForSender) *evictionScoreComputer {
+	computer := &evictionScoreComputer{
+		senders:          senders,
+		scores:           make([]float64, len(senders)),
+		scoresAsPercents: make([]int64, len(senders)),
+
+		minGas:         math.MaxInt64,
+		minSize:        math.MaxInt64,
+		minOrderNumber: math.MaxInt64,
+		minTxCount:     math.MaxInt64,
+	}
+
+	computer.computeBounds()
+	computer.computeRanges()
+	computer.computeScores()
+	computer.convertScoresToPercents()
+
+	return computer
+}
+
+// computeBounds finds the min and max values for the score parameters (gas, size and so on)
+func (computer *evictionScoreComputer) computeBounds() {
+	for _, sender := range computer.senders {
+		gas := sender.totalGas.Get()
+		size := sender.totalBytes.Get()
+		txCount := sender.countTx()
+		orderNumber := sender.orderNumber
+
+		if gas > computer.maxGas {
+			computer.maxGas = gas
+		}
+		if gas < computer.minGas {
+			computer.minGas = gas
+		}
+
+		if size > computer.maxSize {
+			computer.maxSize = size
+		}
+		if size < computer.minSize {
+			computer.minSize = size
+		}
+
+		if txCount > computer.maxTxCount {
+			computer.maxTxCount = txCount
+		}
+		if txCount < computer.minTxCount {
+			computer.minTxCount = txCount
+		}
+
+		if orderNumber > computer.maxOrderNumber {
+			computer.maxOrderNumber = orderNumber
+		}
+		if orderNumber < computer.minOrderNumber {
+			computer.minOrderNumber = orderNumber
+		}
+	}
+}
+
+func (computer *evictionScoreComputer) computeRanges() {
+	computer.gasRange = strictlyPositive(computer.maxGas - computer.minGas)
+	computer.sizeRange = strictlyPositive(computer.maxSize - computer.minSize)
+	computer.orderNumberRange = strictlyPositive(computer.maxOrderNumber - computer.minOrderNumber)
+	computer.txCountRange = strictlyPositive(computer.maxTxCount - computer.minTxCount)
+}
+
+func strictlyPositive(value int64) int64 {
+	if value > 0 {
+		return value
+	}
+	return 1
+}
+
+func (computer *evictionScoreComputer) computeScores() {
+	for i, sender := range computer.senders {
+		computer.scores[i] = computer.computeScore(sender)
+	}
+}
+
+// A low score means that the sender will be evicted
+// A high score means that the sender will not be evicted, most probably
+// Score is:
+// - inversely proportional to sender's tx count
+// - inversely proportional to sender's tx total size
+// - directly proportional to sender's order number
+// - directly proportional to sender's tx total gas
```

Comment:
for the gas did I understand correctly that you return the numberOfGasUnits*gasPricePerUnit?

if yes, then if we have two senders, one offering 20k gasUnits with a price of 10 and another offering 201k gasUnits with a price of 1 , then the second one would have a higher score, even if in the case of normal transaction the first one would have been more advantageous (for second case, the extra gas units would be refunded)

Reply:
Indeed, for the gas we have this formula:

```
func estimateTxGas(tx data.TransactionHandler) int64 {
	gasPrice := int64(tx.GetGasPrice())
	gasLimit := int64(tx.GetGasLimit())
	return gasPrice * gasLimit
}
```

We can probably give a higher coefficient for `gasPrice`, in order to make it more significant for the score (we can also increase its power in a polynomial manner).

Reply:
I've fixed the gas formula. Now we use the average fee (micro ERD per processing unit) when computing the score for a sender.

See `estimateTxFee()`. And:

```
PPUAvg := float64(params.fee) / float64(params.gas)
PPUScore := math.Pow(PPUAvg/PPUMin, 3)
```
Raised at power 3, since `x^3` grows fast, but not that fast as the exponential.